scripts/train_lm.py

```diff
@@ -146,12 +146,39 @@ def WriteMetaparameters(metaparameters, ngram_order, num_train_sets, out_file):
 
 def FormatMetaparameters(metaparameters):
     out = []
+    # get the maximum length of decimals of metaparameters
+    decimal_length = 0
     for param in metaparameters:
-        x = '{:.3f}'.format(param)
-        if x == '0.00':
-            x = '{:.3g}'.format(param)
-        out.append(x)
-
+        temp_length = len(str(param))
+        if decimal_length < temp_length:
+            decimal_length = temp_length
+    decimal_length -= 2
+    
+    temp_original = metaparameters[0]
+    temp = round(temp_original, 3)
+    for param in metaparameters[1:]:
+        round_index = 3
+        x_original = param
+        x = round(x_original, round_index)
+        while(temp == x and round_index <= decimal_length):
```

Comment:
But remember we need to ensure there are no collisions also.

On Tue, Aug 16, 2016 at 8:08 PM, Wang Jian <notifications@github.com> wrote:

> In scripts/train_lm.py
> <https://github.com/danpovey/pocolm/pull/52#discussion_r75056653>:
>
> > -        if x == '0.00':
> > -            x = '{:.3g}'.format(param)
> > -        out.append(x)
> > -
> > +        temp_length = len(str(param))
> > +        if decimal_length < temp_length:
> > +            decimal_length = temp_length
> > +    decimal_length -= 2
> > +
> > +    temp_original = metaparameters[0]
> > +    temp = round(temp_original, 3)
> > +    for param in metaparameters[1:]:
> > +        round_index = 3
> > +        x_original = param
> > +        x = round(x_original, round_index)
> > +        while(temp == x and round_index <= decimal_length):
>
> Hi Ke,
> I think what Dan mean is your logic is a little complicated. We could do
> this inside the for-loop without converting the numbers to string and
> calculating the length.
> Assume we have a function AppendDigitUtilNonZero(f, n), which will append
> n digits after the decimal point except zero. The code could be something
> like this:
>
> out = []
> for param in metaparameters:
>     num_non_zero_digit = 3
>     x = AppendDigitUtilNonZero(param, num_non_zero_digit)
>     # make sure we don't get exact ones or zeros
>     if x == '0.000' or x == '1.000':
>         num_non_zero_digit += 1
>         x = AppendDigitUtilNonZero(param, num_non_zero_digit)
>     # make sure we don't get collisions
>     for s in out:
>         if x == s:
>             num_non_zero_digit += 1
>             x = AppendDigitUtilNonZero(param, num_non_zero_digit)
>     out.append(x)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/danpovey/pocolm/pull/52/files/3bb42c2f1362ffd85d6a4cf0b1fe1a50ab2fe9e1#r75056653>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu0StZCa7gEFBXMl8wy1WwJzNZzwMks5qgnskgaJpZM4Jl6Bt>
> .
>